### PR TITLE
PP-5358 Include GovUkPayEvents in GoCardless mandate state calculation

### PR DESCRIPTION
**Pull request by stephencdaly**
When calculating the state of a GoCardless mandate, find the most recent GoCardlessEvent and GovUkPayEvent then use whichever has a later timestamp to calculate the state.

### CODE REVIEW

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java [hunk 1]
```diff
@@ -32,15 +50,39 @@
...
     Optional<DirectDebitStateWithDetails<MandateState>> calculate(Mandate mandate) {
-        Optional<GoCardlessEvent> latestApplicableGoCardlessEvent = mandate.getPaymentProviderMandateId()
+        Optional<GoCardlessEvent> latestApplicableGoCardlessEvent = getLatestApplicableGoCardlessEvent(mandate);
+
+        Optional<GovUkPayEvent> latestApplicableGovUkPayEvent 
+                = govUkPayEventDao.findLatestApplicableEventForMandate(mandate.getId(), GOV_UK_PAY_EVENT_TYPES_THAT_CHANGE_STATE);
+
+        return Stream.of(latestApplicableGoCardlessEvent, latestApplicableGovUkPayEvent)
+                .flatMap(Optional::stream)
+                .max(Comparator.comparing(Event::getTimestamp))
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: Neat, don’t think I’ve ever seen `max(…)` used in the real world before!

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java [hunk 2]
```diff
@@ -51,14 +93,37 @@
...
+
+    private DirectDebitStateWithDetails<MandateState> mapEventToState(Event event, Mandate mandate) {
+        if (event instanceof GoCardlessEvent) {
+            return mapGoCardlessEventToState((GoCardlessEvent) event, mandate);
+        } else if (event instanceof GovUkPayEvent) {
+            return mapGovUkPayEventToState((GovUkPayEvent) event, mandate);
+        } else {
+            throw new IllegalArgumentException(format("Unexpected Event of type %s", event.getClass()));
+        }
+    }
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: We should consider reintroducing an enum for GoCardless event actions again. Then we could have a common interface for that and `GovUkPayEventType` and use a single map for mapping them to events, avoiding the need for this crime against polymorphism 😛

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java [hunk 3]
```diff
@@ -51,14 +93,37 @@
...
+
+    private DirectDebitStateWithDetails<MandateState> mapGoCardlessEventToState(GoCardlessEvent goCardlessEvent, Mandate mandate) {
+        if (!GOCARDLESS_ACTION_TO_MANDATE_STATE.containsKey(goCardlessEvent.getAction())) {
+            throw new UnexpectedGoCardlessEventActionException(goCardlessEvent.getAction(), mandate);
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: Can this ever reasonably happen given that we got the relevant event based on the key set of the map?

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java [hunk 4]
```diff
@@ -51,14 +93,37 @@
...
-                );
+    private DirectDebitStateWithDetails<MandateState> mapGovUkPayEventToState(GovUkPayEvent govUkPayEvent, Mandate mandate) {
+        if (!GOV_UK_PAY_EVENT_TYPE_MANDATE_STATE.containsKey(govUkPayEvent.getEventType())) {
+            throw new UnexpectedGovUkPayEventTypeException(govUkPayEvent.getEventType(), mandate);
+        }
+        return new DirectDebitStateWithDetails<>(
+                GOV_UK_PAY_EVENT_TYPE_MANDATE_STATE.get(govUkPayEvent.getEventType()),
+                null,
+                null
+        );
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: I wonder if it’s time for an overloaded constructor

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java (hunk 3 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: no, not reasonably, should I just let it throw a NullPointerException if somehow the DAO returns something unexpected?

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java (hunk 4 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: ok

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java (hunk 3 above)
Comment (alexbishop1) [reply src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: I’d say so. It’s still a `RuntimeException`, so the effect will be basically the same.

**Comment by stephencdaly** [comment]
I've added 2 extra commits with changes

File: src/main/java/uk/gov/pay/directdebit/mandate/exception/UnexpectedGoCardlessEventActionException.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.pay.directdebit.mandate.exception;
+
+import uk.gov.pay.directdebit.mandate.model.Mandate;
+
+import static java.lang.String.format;
+
+public class UnexpectedGoCardlessEventActionException extends RuntimeException {
+    public UnexpectedGoCardlessEventActionException(String action, Mandate mandate) {
+        super(format("Unhandled GoCardless event action %s when calculating state for mandate %s", action, mandate.getExternalId()));
```
[the remark this answers is not in the document]
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/mandate/exception/UnexpectedGoCardlessEventActionException.java]: changed

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java [hunk 6]
```diff
@@ -32,15 +50,39 @@
             "failed", FAILED
     );
 
+    private static final Map<GovUkPayEvent.GovUkPayEventType, MandateState> GOV_UK_PAY_EVENT_TYPE_MANDATE_STATE = Map.of(
```
[the remark this answers is not in the document]
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: changed

File: src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java (hunk 2 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java]: Added a method on the `Event` interface so we don't have to do this.

File: src/main/java/uk/gov/pay/directdebit/events/model/GoCardlessEvent.java [hunk 7]
```diff
@@ -1,15 +1,35 @@
...
 
 public class GoCardlessEvent implements Event {
 
+    private static final Map<String, MandateState> GOCARDLESS_ACTION_TO_MANDATE_STATE = Map.of(
+            "created", CREATED,
+            "submitted", SUBMITTED,
+            "active", ACTIVE,
+            "cancelled", CANCELLED,
+            "failed", FAILED
+    );
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/events/model/GoCardlessEvent.java]: I’m not too keen on having the states an event maps to be part of the event. It feels like conflating the event with how we interpret the event and ties us quite tightly to the idea that each event maps to exactly one state.

Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/events/model/GoCardlessEvent.java]: should I just go back to having the instanceof stuff then?

Comment (alexbishop1) [reply src/main/java/uk/gov/pay/directdebit/events/model/GoCardlessEvent.java]: Yeah, for now

**Pull request merged**